File: westcoast/src/main/java/frc/robot/definition/Controllers.java
package frc.robot.definition;

import com.revrobotics.SparkMaxPIDController;
import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.controller.ProfiledPIDController;
import org.jetbrains.annotations.NotNull;

public interface Controllers {
    @NotNull
    PIDController gyroDrive();

    @NotNull
    ProfiledPIDController leaveLine();

    @NotNull
    ProfiledPIDController perfectTurn();

    @NotNull
    PIDController alignToTarget();

    @NotNull
    PIDController stayOnTarget();
    
    @NotNull
    PIDController forwardAlignToTarget();

    void shooterConfig(@NotNull SparkMaxPIDController controller);

    void climberConfigDeploy(@NotNull SparkMaxPIDController controller);

    void climberConfigClimb(@NotNull SparkMaxPIDController controller);

    interface ControllerConfigurator {
        void configure(@NotNull SparkMaxPIDController controller);
    }
}

File: westcoast/src/main/kotlin/frc/robot/definition/Controllers.kt
package frc.robot.definition

import com.revrobotics.SparkMaxPIDController
import edu.wpi.first.math.controller.PIDController
import edu.wpi.first.math.controller.ProfiledPIDController

class Controllers(
    val leaveLine: () -> ProfiledPIDController,
    val setupShotTurn: () -> PIDController,
    val setupShotForward: () -> PIDController,
    val shooterConfig: SparkMaxPIDController.() -> Unit,
    val climber: Climber
) {
    data class Climber(
        val deploy: SparkMaxPIDController.() -> Unit,
        val climb: SparkMaxPIDController.() -> Unit
    )
}